tools/cndump/kvset_dump.c

```diff
@@ -0,0 +1,965 @@
+        omf_hbh_vgmap_off_pg(hbh), omf_hbh_vgmap_len_pg(hbh),
+        omf_hbh_hlog_off_pg(hbh), omf_hbh_hlog_len_pg(hbh));
+
+    printf("  max_ptomb byte off/len %u %u, min_ptomb off/len %u %u\n",
+        omf_hbh_max_pfx_off(hbh),
+        omf_hbh_max_pfx_len(hbh),
+        omf_hbh_min_pfx_off(hbh),
+        omf_hbh_min_pfx_len(hbh));
+
+    /* TODO: dump the ptomb tree */
+}
+
+static void
+kblock_dump(struct dump_mblock *mblk)
+{
+    const void *p = mblk->data;;
+    const struct kblock_hdr_omf *kbh = p;
+    const struct wbt_hdr_omf *   wbt = p + omf_kbh_wbt_hoff(kbh);
+    const struct bloom_hdr_omf  *bh = p + omf_kbh_blm_hoff(kbh);
+
+    void *buf = NULL;
+    size_t bufsz = 0;
+
+    printf("\nkblock: ");
+    dump_props(&mblk->props);
+    printf("  magic 0x%08x  ver %d  nkey %d  ntomb %d\n",
+        omf_kbh_magic(kbh), omf_kbh_version(kbh), omf_kbh_entries(kbh), omf_kbh_tombs(kbh));
+
+    printf("  metrics: keys %u tombs %u key_bytes %u val_bytes %lu\n",
+        omf_kbh_entries(kbh),
+        omf_kbh_tombs(kbh),
+        omf_kbh_key_bytes(kbh),
+        omf_kbh_val_bytes(kbh));
+
+    printf("  wbt: hdr %d %d  data_pg %d %d  ver %u\n",
+        omf_kbh_wbt_hoff(kbh),
+        omf_kbh_wbt_hlen(kbh),
+        omf_kbh_wbt_doff_pg(kbh),
+        omf_kbh_wbt_dlen_pg(kbh),
+        omf_wbt_version(wbt));
+
+    printf("  bloom: hdr %d %d  data_pg %d %d  ver %u\n",
+        omf_kbh_blm_hoff(kbh),
+        omf_kbh_blm_hlen(kbh),
+        omf_kbh_blm_doff_pg(kbh),
+        omf_kbh_blm_dlen_pg(kbh),
+        omf_bh_version(bh));
+
+    printf("  kmd: start_pg %u\n",
+        omf_kbh_wbt_doff_pg(kbh) + omf_wbt_root(wbt) + 1);
+
+    printf("  keymin: off %u len %u key %s\n",
+        omf_kbh_min_koff(kbh),
+        omf_kbh_min_klen(kbh),
+        fmt_data(&buf, &bufsz, p + omf_kbh_min_koff(kbh), omf_kbh_min_klen(kbh),
+            opts.min_max_klen, opts.penc));
+
+    printf("  keymax: off %u len %u key %s\n",
+        omf_kbh_max_koff(kbh),
+        omf_kbh_max_klen(kbh),
+        fmt_data(&buf, &bufsz, p + omf_kbh_max_koff(kbh), omf_kbh_max_klen(kbh),
+            opts.min_max_klen, opts.penc));
+
+    bloom_dump(mblk);
+    wbt_dump(mblk);
+    free(buf);
+}
+
+static void
+vblock_dump(struct dump_mblock *mblk)
+{
+    const struct vblock_footer_omf *vbf;
+    uint min_koff, min_klen, max_koff, max_klen;
+    uint32_t magic, version;
+    uint64_t vgroup;
+    void *buf = NULL;
+    size_t bufsz = 0;
+
+    vbf = mblk->data + mblk->props.mpr_write_len - VBLOCK_FOOTER_LEN;
+
+    printf("vblock: ");
+    dump_props(&mblk->props);
+
+    magic = omf_vbf_magic(vbf);
+    version = omf_vbf_version(vbf);
+    printf("  magic 0x%x vers %u\n", magic, version);
+
+    if (magic != VBLOCK_FOOTER_MAGIC) {
+        printf("  BAD MAGIC (expected 0x%x)\n", VBLOCK_FOOTER_MAGIC);
+        return;
+    }
+
+    if (version != VBLOCK_FOOTER_VERSION) {
+        printf("  unsupported version (expected %u)\n", VBLOCK_FOOTER_VERSION);
```

Comment:
all caps like the previous message so it stands out?

Reply:
magic seemed more severe than version.

Reply:
You should print the version you got as well, maddening otherwise..